File: App/NodeProcessors/SortNode.jl
module SortNode

using Dates, DataFrames
using ..VectoriseBacktestService
using ..VectoriseBacktestService.Types
using ..VectoriseBacktestService.FlowData
using ..VectoriseBacktestService.TASorting
using ..VectoriseBacktestService.ErrorHandlers
using ..VectoriseBacktestService.BacktestUtilites
using ..VectoriseBacktestService.ReturnCalculations
using ..VectoriseBacktestService.TimeCalculation

export process_sort_node,
    process_branches, calculate_selection_indices, update_portfolio_history

function update_portfolio_history(
    portfolio_history::Vector{DayData},
    temp_portfolio_vectors::Vector{Vector{DayData}},
    selection_indices::Vector{Vector{Int}},
    daily_weight::Float32,
    common_data_span::Int,
    sort_window::Int,
    selection_count::Int,
)::Nothing
    try
        if isempty(portfolio_history) ||
            isempty(temp_portfolio_vectors) ||
            isempty(selection_indices)
            throw(
                PortfolioUpdateError(
                    "Empty input arrays",
                    Dict(
                        "portfolio_history_length" => length(portfolio_history),
                        "temp_vectors_length" => length(temp_portfolio_vectors),
                        "selection_indices_length" => length(selection_indices),
                    ),
                ),
            )
        end

        for day in 1:(common_data_span)
            try
                for branch_index in selection_indices[day]
                    if branch_index > length(temp_portfolio_vectors)
                        throw(
                            PortfolioUpdateError(
                                "Branch index out of bounds",
                                Dict(
                                    "branch_index" => branch_index,
                                    "available_branches" => length(temp_portfolio_vectors),
                                ),
                            ),
                        )
                    end

                    for stock in temp_portfolio_vectors[branch_index][day].stockList
                        stock.weightTomorrow /= selection_count
                        stock.weightTomorrow *= daily_weight
                        push!(
                            portfolio_history[end - (common_data_span) + day].stockList,
                            stock,
                        )
                    end
                end
            catch e
                throw(
                    PortfolioUpdateError(
                        "Failed to process day $day", Dict("day" => day, "error" => e)
                    ),
                )
            end
        end
        return nothing
    catch e
        if e isa BacktestError
            rethrow(e)
        else
            throw(PortfolioUpdateError("Unexpected error in portfolio update", e))
        end
    end
end

function calculate_selection_indices(
    branch_metrics::Vector{Vector{Float64}},
    active_branch_mask::BitVector,
    data_span::Int,
    sort_window::Int,
    select_function::String,
    selection_count::Int,
    sort_node::Dict{String,Any},
    flow_count::Dict{String,Int},
)::Vector{Vector{Int}}
    try
        if isempty(branch_metrics)
            throw(SelectionError("Empty branch metrics", nothing))
        end

        if selection_count > length(branch_metrics)
            throw(
                SelectionError(
                    "Selection count exceeds available branches",
                    Dict(
                        "selection_count" => selection_count,
                        "available_branches" => length(branch_metrics),
                    ),
                ),
            )
        end

        selection_indices = [Int[] for _ in 1:(data_span)]
        active_days = findall(active_branch_mask[(end - (data_span) + 1):end])

        for day in active_days
            if haskey(sort_node, "hash")
                increment_flow_count(flow_count, sort_node["hash"])
            end

            day_metrics = [metric[day] for metric in branch_metrics]
            if any(isnan, day_metrics) || any(isinf, day_metrics)
                # replace NaN and Inf with 0
                day_metrics = [
                    if isnan(metric) || isinf(metric)
                        0.0f0
                    else
                        metric
                    end for metric in day_metrics
                ]
            end

            selection_indices[day] = if select_function == "Top"
                partialsortperm(day_metrics, 1:selection_count; rev=true)
            elseif select_function == "Bottom"
                partialsortperm(day_metrics, 1:selection_count; rev=false)
            else
                throw(SelectionError("Invalid select function", select_function))
            end
        end

        return selection_indices
    catch e
        if e isa BacktestError
            rethrow(e)
        else
            throw(SelectionError("Failed to calculate selection indices", e))
        end
    end
end
function process_branches(
    branches::Dict{String,Any},
    branch_keys::Vector{String},
    total_days::Int,
    daily_weight::Float32,
    date_range::Vector{String},
    end_date::Date,
    flow_count::Dict{String,Int},
    flow_stocks::Dict{String,Vector{DayData}},
    indicator_cache::Dict{String,Vector{Float32}},
    price_cache::Dict{String,DataFrame},
    strategy_root::Dict{String,Any},
    live_execution::Bool,
    global_cache_length::Int,
)::Tuple{Vector{Vector{DayData}},Int}
    try
        if isempty(branches) || isempty(branch_keys)
            throw(
                BranchProcessingError(
                    "Empty branches or branch keys",
                    Dict(
                        "branches_count" => length(branches),
                        "branch_keys_count" => length(branch_keys),
                    ),
                ),
            )
        end

        temp_portfolio_vectors = [
            [DayData() for _ in 1:total_days] for _ in 1:length(branch_keys)
        ]
        min_data_length = total_days

        for (current_branch_idx, branch_key) in enumerate(branch_keys)
            if !haskey(branches, branch_key)
                throw(BranchProcessingError("Branch key not found", branch_key))
            end

            if isempty(branches[branch_key])
                throw(BranchProcessingError("Empty branch", branch_key))
            end

            temp_active_branch_mask = BitVector(trues(total_days))
            branch_data_length = post_order_dfs(
                branches[branch_key][1],
                temp_active_branch_mask,
                total_days,
                1.0f0,
                temp_portfolio_vectors[current_branch_idx],
                date_range,
                end_date,
                flow_count,
                flow_stocks,
                indicator_cache,
                price_cache,
                global_cache_length,
                strategy_root,
                live_execution,
            )
            min_data_length = min(min_data_length, branch_data_length)
        end

        return temp_portfolio_vectors, min_data_length
    catch e
        if e isa BacktestError
            rethrow(e)
        else
            throw(BranchProcessingError("Failed to process branches", e))
        end
    end
end
function validate_sort_node(
    sort_node::Dict{String,Any}, portfolio_history::Vector{DayData}, total_days::Int
)
    if !haskey(sort_node, "branches")
        throw(ValidationError("Sort node missing 'branches' field"))
    end

    if isempty(portfolio_history) || total_days <= 0
        throw(
            ValidationError(
                "Invalid input parameters",
                Dict(
                    "portfolio_history_length" => length(portfolio_history),
                    "total_days" => total_days,
                ),
            ),
        )
    end
end

function process_sort_node(
    sort_node::Dict{String,Any},
    active_branch_mask::BitVector,
    total_days::Int,
    daily_weight::Float32,
    portfolio_history::Vector{DayData},
    date_range::Vector{String},
    end_date::Date,
    flow_count::Dict{String,Int},
    flow_stocks::Dict{String,Vector{DayData}},
    indicator_cache::Dict{String,Vector{Float32}},
    price_cache::Dict{String,DataFrame},
    strategy_root::Dict{String,Any},
    live_execution::Bool,
    global_cache_length::Int,
)::Int
    node_id = get(sort_node, "id", "unknown")
    # Initial validation
    validate_sort_node(sort_node, portfolio_history, total_days)
    # Get branches and properties in one block
    local branches, has_folder_node, selection_count, select_function, sort_function, sort_window
    branches, has_folder_node = get_branches(sort_node)
    isempty(branches) && throw(ValidationError("Sort node has no branches"))

    selection_count, select_function = parse_select_properties(sort_node)
    selection_count <= 0 &&
        throw(ValidationError("Invalid selection count", selection_count))

    sort_function, sort_window = parse_sort_properties(sort_node)

    # Process branches
    branch_keys = collect(keys(branches))
    branch_metrics = Vector{Vector{Float32}}(undef, length(branch_keys))

    # Process branches and calculate metrics
    local temp_portfolio_vectors, common_data_span, flag, original_total_days
    flag = false
    original_total_days = total_days
    uses_delta_in_sort_function::Bool =
        sort_function in ["Standard Deviation of Return", "Moving Average of Return"]
    pad_252_days =
        sort_function in ["Relative Strength Index", "Exponential Moving Average of price"]
    """
        Refer to AllocationNode.jl for a detailed explanation of this if-else block as well as cached_mapping/cached_portfolio_values.

        The variable 'uses_delta_in_sort_function' is utilized when the sort function computes the deltas
        of the returns. In such cases, the total days should be incremented by the sort window and an additional day
        to ensure sufficient data is available for calculating the branch_return_curves and subsequently the branch_metrics.
    """
    try
        new_date_range = date_range
        if has_folder_node
            total_days += sort_window + uses_delta_in_sort_function + pad_252_days * 252
            new_date_range = populate_dates(total_days, end_date, String[], live_execution)
            total_days = min(total_days, length(new_date_range))
            new_active_branch_mask = BitVector(trues(total_days))
            for day in eachindex(active_branch_mask)
                new_active_branch_mask[(end - day + 1)] = active_branch_mask[(end - day + 1)]
            end
            active_branch_mask = new_active_branch_mask
            flag = true
        end
        temp_portfolio_vectors, common_data_span = process_branches(
            branches,
            branch_keys,
            total_days,
            daily_weight,
            new_date_range,
            end_date,
            flow_count,
            flow_stocks,
            indicator_cache,
            price_cache,
            strategy_root,
            live_execution,
            global_cache_length,
        )
        if has_folder_node
            folder_min_data_length::Int = total_days
            cached_portfolio_values::Dict{String,Vector{Float64}}, use::Bool, cached_end_date::Union{Date,Nothing} = read_portfolio_values(
                sort_node["nodeChildrenHash"]
            )
            cached_mapping::Dict{String,Float64} = Dict{String,Float64}()
            branch_return_curves::Vector{Vector{Float64}} = Vector{Vector{Float64}}()
            if use && (flag) && cached_end_date >= end_date
                for (key, value) in cached_portfolio_values
                    # Push the current array to the 2D vector
                    push!(branch_return_curves, value)
                    folder_min_data_length = min(folder_min_data_length, length(value))
                end
                common_data_span = min(common_data_span, folder_min_data_length)
                branch_metrics = apply_sort_function(
                    branch_return_curves, sort_function, sort_window
                )
            else
                if use && (flag)
                    for (key, value) in cached_portfolio_values
                        cached_mapping[key] = value[max(1, end - original_total_days + 1)]
                    end
                end

                branch_return_curves, folder_min_data_length = calculate_branch_return_curves(
                    temp_portfolio_vectors,
                    new_date_range[max(1, (end - total_days + 1)):end],
                    end_date,
                    price_cache,
                    total_days,
                    sort_function,
                    cached_mapping,
                    live_execution,
                )
                fresh_portfolio_values = Dict{String,Vector{Float64}}()
                for (i, ret) in enumerate(branch_return_curves)
                    append_val = ret .- 100.0f0
                    if haskey(cached_portfolio_values, string(i))
                        fresh_portfolio_values[string(i)] = append!(
                            cached_portfolio_values[string(i)],
                            append_val[max(1, (end - original_total_days + 1)):end],
                        )
                    else
                        fresh_portfolio_values[string(i)] = append_val[max(
                            1, (end - original_total_days + 1)
                        ):end]
                    end
                end
                cache_portfolio_values(
                    fresh_portfolio_values,
                    sort_node["nodeChildrenHash"],
                    Date(end_date),
                    live_execution,
                )

                common_data_span = min(folder_min_data_length, common_data_span)
                branch_metrics = apply_sort_function(
                    branch_return_curves, sort_function, sort_window
                )
            end
        else
            branch_metrics, stock_min_data_length = apply_sort_ta_stock_function(
                temp_portfolio_vectors,
                date_range,
                end_date,
                price_cache,
                total_days,
                sort_function,
                sort_window,
                indicator_cache,
                live_execution,
            )
            common_data_span = min(stock_min_data_length, common_data_span)
        end
    catch e
        throw(
            ProcessingError(
                "Failed to process branches and metrics",
                Dict("node_id" => node_id, "error" => e),
            ),
        )
    end

    # Adjust data spans and calculate final results
    try
        # Adjust branch metrics
        # Flag indicates N-day backtest, in which case the data_span is the padded total_days
        data_span = if has_folder_node
            (common_data_span - pad_252_days * 252) - sort_window
        else
            common_data_span
        end
        # If the sort function uses deltas, the data span should be decremented by 1 in order to adjust the 1 we added earlier
        data_span -= (uses_delta_in_sort_function && flag)
        if data_span < 0
            data_span = if has_folder_node
                (total_days - pad_252_days * 252) - sort_window -
                (uses_delta_in_sort_function && flag)
            else
                common_data_span
            end
        end

        branch_metrics = [
            metric[max(1, (end - data_span + 1)):end] for metric in branch_metrics
        ]
        data_span = min(data_span, length(branch_metrics[1]))
        # Calculate selection indices
        selection_indices = calculate_selection_indices(
            branch_metrics,
            active_branch_mask,
            data_span,
            sort_window,
            select_function,
            selection_count,
            sort_node,
            flow_count,
        )
        # Adjust portfolio vectors
        temp_portfolio_vectors = [
            days[(end - data_span + 1):end] for days in temp_portfolio_vectors
        ]

        # Update portfolio history
        update_portfolio_history(
            portfolio_history,
            temp_portfolio_vectors,
            selection_indices,
            daily_weight,
            data_span,
            sort_window,
            selection_count,
        )
        # Set flow stocks if needed
        if haskey(sort_node, "hash")
            set_flow_stocks(flow_stocks, portfolio_history, sort_node["hash"])
        end
        return if flag
            original_total_days
        else
            if has_folder_node
                min((total_days - sort_window), (common_data_span - sort_window))
            else
                min(common_data_span, total_days)
            end
        end
    catch e
        throw(
            ProcessingError(
                "Failed to finalize sort node processing",
                Dict("node_id" => node_id, "error" => e),
            ),
        )
    end
end

end

File: cpp_backtester/src/nodes/sort_node.cpp
#include \"sort_node.h\"
#include <algorithm>
#include <numeric>
#include <cmath>
#include <stdexcept>

namespace atlas {

NodeResult SortNodeProcessor::process(
    const StrategyNode& node,
    std::vector<bool>& active_mask,
    int total_days,
    float node_weight,
    std::vector<DayData>& portfolio_history,
    const std::vector<std::string>& date_range,
    std::unordered_map<std::string, int>& flow_count,
    std::unordered_map<std::string, std::vector<DayData>>& flow_stocks,
    std::unordered_map<std::string, std::vector<float>>& indicator_cache,
    std::unordered_map<std::string, std::vector<float>>& price_cache,
    const Strategy& strategy,
    bool live_execution,
    int global_cache_length
) {
    try {
        // Validate sort node
        if (!validate_sort_node(node)) {
            return NodeResult(0, false, \"Invalid sort node structure\");
        }
        
        // Parse properties
        SelectFunction select_function;
        int selection_count;
        if (!parse_select_properties(node.properties, select_function, selection_count)) {
            return NodeResult(0, false, \"Invalid selection properties\");
        }
        
        SortFunction sort_function;
        int sort_window;
        if (!parse_sort_properties(node.properties, sort_function, sort_window)) {
            return NodeResult(0, false, \"Invalid sort properties\");
        }
        
        // Get branches
        auto [branches, has_folder_node] = get_branches(node);
        if (branches.empty()) {
            return NodeResult(0, false, \"Sort node has no branches\");
        }
        
        // Get branch keys
        std::vector<std::string> branch_keys;
        for (auto it = branches.begin(); it != branches.end(); ++it) {
            branch_keys.push_back(it.key());
        }
        
        if (selection_count > static_cast<int>(branch_keys.size())) {
            return NodeResult(0, false, \"Selection count exceeds available branches\");
        }
        
        // Adjust total_days based on sort function requirements
        int original_total_days = total_days;
        bool uses_delta = (sort_function == SortFunction::STANDARD_DEVIATION_RETURN ||
                          sort_function == SortFunction::MOVING_AVERAGE_RETURN);
        bool pad_252_days = (sort_function == SortFunction::RELATIVE_STRENGTH_INDEX ||
                             sort_function == SortFunction::EXPONENTIAL_MOVING_AVERAGE);
        
        if (has_folder_node) {
            total_days += sort_window + (uses_delta ? 1 : 0) + (pad_252_days ? 252 : 0);
            // Update active mask accordingly
            std::vector<bool> new_active_mask(total_days, true);
            int offset = std::max(0, static_cast<int>(new_active_mask.size()) - static_cast<int>(active_mask.size()));
            for (size_t i = 0; i < active_mask.size(); ++i) {
                size_t new_idx = offset + i;
                if (new_idx < new_active_mask.size()) {
                    new_active_mask[new_idx] = active_mask[i];
                }
            }
            active_mask = new_active_mask;
        }
        
        // Process branches
        auto [temp_portfolio_vectors, common_data_span] = process_branches(
            branches, branch_keys, total_days, node_weight, date_range,
            flow_count, flow_stocks, indicator_cache, price_cache,
            strategy, live_execution, global_cache_length
        );
        
        // Calculate metrics for sorting
        auto branch_metrics = calculate_branch_metrics(
            temp_portfolio_vectors, date_range, sort_function, sort_window,
            indicator_cache, price_cache, live_execution
        );
        
        // Calculate selection indices
        auto selection_indices = calculate_selection_indices(
            branch_metrics, active_mask, common_data_span, select_function,
            selection_count, node, flow_count
        );
        
        // Update portfolio history
        update_portfolio_history(
            portfolio_history, temp_portfolio_vectors, selection_indices,
            node_weight, common_data_span, selection_count
        );
        
        // Set flow stocks
        if (!node.hash.empty()) {
            set_flow_stocks(flow_stocks, portfolio_history, node.hash);
        }
        
        return NodeResult(common_data_span, true);
        
    } catch (const std::exception& e) {
        return NodeResult(0, false, \"Sort node error: \" + std::string(e.what()));
    }
}

bool SortNodeProcessor::validate_sort_node(const StrategyNode& node) const {
    // Must have branches
    if (!node.branches.contains(\"branches\") && node.branches.empty()) {
        return false;
    }
    
    // Must have select and sortby properties
    return node.properties.contains(\"select\") && node.properties.contains(\"sortby\");
}

bool SortNodeProcessor::parse_select_properties(
    const nlohmann::json& properties,
    SelectFunction& select_function,
    int& selection_count
) const {
    if (!properties.contains(\"select\")) {
        return false;
    }
    
    auto select_obj = properties[\"select\"];
    if (!select_obj.contains(\"function\") || !select_obj.contains(\"howmany\")) {
        return false;
    }
    
    // Parse select function
    std::string function_str = select_obj[\"function\"].get<std::string>();
    try {
        select_function = parse_select_function(function_str);
    } catch (...) {
        return false;
    }
    
    // Parse selection count
    try {
        selection_count = std::stoi(select_obj[\"howmany\"].get<std::string>());
    } catch (...) {
        return false;
    }
    
    return selection_count > 0;
}

bool SortNodeProcessor::parse_sort_properties(
    const nlohmann::json& properties,
    SortFunction& sort_function,
    int& sort_window
) const {
    if (!properties.contains(\"sortby\")) {
        return false;
    }
    
    auto sortby_obj = properties[\"sortby\"];
    if (!sortby_obj.contains(\"function\")) {
        return false;
    }
    
    // Parse sort function
    std::string function_str = sortby_obj[\"function\"].get<std::string>();
    try {
        sort_function = parse_sort_function(function_str);
    } catch (...) {
        return false;
    }
    
    // Parse sort window (default to 20 if not specified)
    sort_window = 20;
    if (sortby_obj.contains(\"window\")) {
        try {
            sort_window = std::stoi(sortby_obj[\"window\"].get<std::string>());
        } catch (...) {
            // Use default
        }
    } else if (sortby_obj.contains(\"period\")) {
        try {
            sort_window = std::stoi(sortby_obj[\"period\"].get<std::string>());
        } catch (...) {
            // Use default
        }
    }
    
    return sort_window > 0;
}

std::tuple<std::vector<std::vector<DayData>>, int> SortNodeProcessor::process_branches(
    const nlohmann::json& branches,
    const std::vector<std::string>& branch_keys,
    int total_days,
    float node_weight,
    const std::vector<std::string>& date_range,
    std::unordered_map<std::string, int>& flow_count,
    std::unordered_map<std::string, std::vector<DayData>>& flow_stocks,
    std::unordered_map<std::string, std::vector<float>>& indicator_cache,
    std::unordered_map<std::string, std::vector<float>>& price_cache,
    const Strategy& strategy,
    bool live_execution,
    int global_cache_length
) {
    std::vector<std::vector<DayData>> temp_portfolio_vectors;
    temp_portfolio_vectors.reserve(branch_keys.size());
    
    int min_data_length = total_days;
    
    for (const auto& branch_key : branch_keys) {
        if (!branches.contains(branch_key)) {
            throw SortNodeError(\"Branch key not found: \" + branch_key);
        }
        
        auto branch_nodes = branches[branch_key];
        if (!branch_nodes.is_array() || branch_nodes.empty()) {
            throw SortNodeError(\"Empty branch: \" + branch_key);
        }
        
        // Initialize portfolio for this branch
        std::vector<DayData> branch_portfolio(total_days);
        
        // For simplicity, we'll process the first node in each branch
        // In a complete implementation, this would process all nodes in the branch
        auto first_node = StrategyNode(branch_nodes[0]);
        
        // Process stock nodes directly (simplified implementation)
        if (first_node.type == \"stock\" && first_node.properties.contains(\"symbol\")) {
            std::string symbol = first_node.properties[\"symbol\"].get<std::string>();
            
            // Add stock to all days for this branch
            for (auto& day : branch_portfolio) {
                day.add_stock(StockInfo(symbol, 1.0f));
            }
        }
        
        temp_portfolio_vectors.push_back(std::move(branch_portfolio));
        // For now, we don't reduce min_data_length since we're using simplified processing
    }
    
    return std::make_tuple(temp_portfolio_vectors, min_data_length);
}

std::vector<std::vector<float>> SortNodeProcessor::calculate_branch_metrics(
    const std::vector<std::vector<DayData>>& temp_portfolio_vectors,
    const std::vector<std::string>& date_range,
    SortFunction sort_function,
    int sort_window,
    std::unordered_map<std::string, std::vector<float>>& indicator_cache,
    std::unordered_map<std::string, std::vector<float>>& price_cache,
    bool live_execution
) {
    std::vector<std::vector<float>> branch_metrics;
    branch_metrics.reserve(temp_portfolio_vectors.size());
    
    for (const auto& branch_portfolio : temp_portfolio_vectors) {
        std::vector<float> metrics;
        
        switch (sort_function) {
            case SortFunction::RELATIVE_STRENGTH_INDEX: {
                // Calculate RSI based on portfolio returns
                auto returns = calculate_portfolio_returns(branch_portfolio, price_cache);
                metrics = calculate_rsi(returns, sort_window);
                break;
            }
            case SortFunction::SIMPLE_MOVING_AVERAGE: {
                auto returns = calculate_portfolio_returns(branch_portfolio, price_cache);
                metrics = calculate_sma(returns, sort_window);
                break;
            }
            case SortFunction::EXPONENTIAL_MOVING_AVERAGE: {
                auto returns = calculate_portfolio_returns(branch_portfolio, price_cache);
                metrics = calculate_ema(returns, sort_window);
                break;
            }
            case SortFunction::PORTFOLIO_RETURN: {
                metrics = calculate_portfolio_returns(branch_portfolio, price_cache);
                break;
            }
            default: {
                // For unsupported functions, use dummy values
                metrics.resize(branch_portfolio.size(), 0.0f);
                for (size_t i = 0; i < metrics.size(); ++i) {
                    metrics[i] = static_cast<float>(i) * 0.1f; // Dummy progressive values
                }
                break;
            }
        }
        
        // Ensure metrics vector has the right size
        if (metrics.size() != branch_portfolio.size()) {
            metrics.resize(branch_portfolio.size(), 0.0f);
        }
        
        branch_metrics.push_back(metrics);
    }
    
    return branch_metrics;
}

std::vector<std::vector<int>> SortNodeProcessor::calculate_selection_indices(
    const std::vector<std::vector<float>>& branch_metrics,
    const std::vector<bool>& active_mask,
    int data_span,
    SelectFunction select_function,
    int selection_count,
    const StrategyNode& node,
    std::unordered_map<std::string, int>& flow_count
) {
    std::vector<std::vector<int>> selection_indices(data_span);
    
    // Find active days
    std::vector<int> active_days;
    int start_offset = std::max(0, static_cast<int>(active_mask.size()) - data_span);
    for (int i = 0; i < data_span; ++i) {
        int mask_idx = start_offset + i;
        if (mask_idx < static_cast<int>(active_mask.size()) && active_mask[mask_idx]) {
            active_days.push_back(i);
        }
    }
    
    for (int day : active_days) {
        // Increment flow count
        if (!node.hash.empty()) {
            increment_flow_count(flow_count, node.hash);
        }
        
        // Get metrics for this day from all branches
        std::vector<std::pair<float, int>> day_metrics;
        for (size_t branch_idx = 0; branch_idx < branch_metrics.size(); ++branch_idx) {
            if (day < static_cast<int>(branch_metrics[branch_idx].size())) {
                float metric = branch_metrics[branch_idx][day];
                // Handle NaN and Inf
                if (std::isnan(metric) || std::isinf(metric)) {
                    metric = 0.0f;
                }
                day_metrics.emplace_back(metric, static_cast<int>(branch_idx));
            }
        }
        
        // Sort based on select function
        if (select_function == SelectFunction::TOP) {
            // Sort in descending order (highest first)
            std::sort(day_metrics.begin(), day_metrics.end(), 
                     [](const auto& a, const auto& b) { return a.first > b.first; });
        } else {
            // Sort in ascending order (lowest first)
            std::sort(day_metrics.begin(), day_metrics.end(), 
                     [](const auto& a, const auto& b) { return a.first < b.first; });
        }
        
        // Select top selection_count indices
        std::vector<int> selected_indices;
        int count = std::min(selection_count, static_cast<int>(day_metrics.size()));
        for (int i = 0; i < count; ++i) {
            selected_indices.push_back(day_metrics[i].second);
        }
        
        selection_indices[day] = selected_indices;
    }
    
    return selection_indices;
}

void SortNodeProcessor::update_portfolio_history(
    std::vector<DayData>& portfolio_history,
    const std::vector<std::vector<DayData>>& temp_portfolio_vectors,
    const std::vector<std::vector<int>>& selection_indices,
    float node_weight,
    int common_data_span,
    int selection_count
) {
    for (int day = 0; day < common_data_span; ++day) {
        if (day >= static_cast<int>(selection_indices.size())) {
            continue;
        }
        
        int portfolio_idx = static_cast<int>(portfolio_history.size()) - common_data_span + day;
        if (portfolio_idx < 0 || portfolio_idx >= static_cast<int>(portfolio_history.size())) {
            continue;
        }
        
        for (int branch_idx : selection_indices[day]) {
            if (branch_idx >= static_cast<int>(temp_portfolio_vectors.size()) ||
                day >= static_cast<int>(temp_portfolio_vectors[branch_idx].size())) {
                continue;
            }
            
            const auto& branch_day = temp_portfolio_vectors[branch_idx][day];
            for (const auto& stock : branch_day.stock_list()) {
                // Adjust weight
                float adjusted_weight = stock.weight_tomorrow() / selection_count * node_weight;
                portfolio_history[portfolio_idx].add_stock(StockInfo(stock.ticker(), adjusted_weight));
            }
        }
    }
}

// Utility function implementations
std::vector<float> SortNodeProcessor::calculate_rsi(const std::vector<float>& prices, int period) {
    std::vector<float> rsi_values;
    if (prices.size() < static_cast<size_t>(period + 1)) {
        rsi_values.resize(prices.size(), 50.0f); // Default RSI
        return rsi_values;
    }
    
    std::vector<float> gains, losses;
    
    // Calculate price changes
    for (size_t i = 1; i < prices.size(); ++i) {
        float change = prices[i] - prices[i-1];
        gains.push_back(change > 0 ? change : 0);
        losses.push_back(change < 0 ? -change : 0);
    }
    
    // Calculate initial averages
    float avg_gain = 0, avg_loss = 0;
    for (int i = 0; i < period; ++i) {
        avg_gain += gains[i];
        avg_loss += losses[i];
    }
    avg_gain /= period;
    avg_loss /= period;
    
    // Calculate RSI
    rsi_values.resize(prices.size(), 50.0f);
    
    for (size_t i = period; i < gains.size(); ++i) {
        avg_gain = (avg_gain * (period - 1) + gains[i]) / period;
        avg_loss = (avg_loss * (period - 1) + losses[i]) / period;
        
        float rs = (avg_loss == 0) ? 100 : avg_gain / avg_loss;
        float rsi = 100 - (100 / (1 + rs));
        
        rsi_values[i + 1] = rsi;
    }
    
    return rsi_values;
}

std::vector<float> SortNodeProcessor::calculate_sma(const std::vector<float>& values, int period) {
    std::vector<float> sma_values;
    
    for (size_t i = 0; i < values.size(); ++i) {
        if (i < static_cast<size_t>(period - 1)) {
            sma_values.push_back(values[i]); // Not enough data, use original value
        } else {
            float sum = 0;
            for (int j = 0; j < period; ++j) {
                sum += values[i - j];
            }
            sma_values.push_back(sum / period);
        }
    }
    
    return sma_values;
}

std::vector<float> SortNodeProcessor::calculate_ema(const std::vector<float>& values, int period) {
    std::vector<float> ema_values;
    if (values.empty()) return ema_values;
    
    float multiplier = 2.0f / (period + 1);
    ema_values.push_back(values[0]); // First value is the seed
    
    for (size_t i = 1; i < values.size(); ++i) {
        float ema = (values[i] * multiplier) + (ema_values[i-1] * (1 - multiplier));
        ema_values.push_back(ema);
    }
    
    return ema_values;
}

std::vector<float> SortNodeProcessor::calculate_portfolio_returns(
    const std::vector<DayData>& portfolio_data,
    std::unordered_map<std::string, std::vector<float>>& price_cache
) {
    std::vector<float> returns;
    returns.reserve(portfolio_data.size());
    
    for (size_t day = 0; day < portfolio_data.size(); ++day) {
        float daily_return = 0.0f;
        
        for (const auto& stock : portfolio_data[day].stock_list()) {
            // Get price for this stock (simplified - using dummy price)
            float price = 100.0f + day * 0.1f; // Dummy progressive price
            daily_return += stock.weight_tomorrow() * price;
        }
        
        returns.push_back(daily_return);
    }
    
    return returns;
}

SortFunction SortNodeProcessor::parse_sort_function(const std::string& sort_function_str) {
    if (sort_function_str == \"Relative Strength Index\") return SortFunction::RELATIVE_STRENGTH_INDEX;
    if (sort_function_str == \"Simple Moving Average of Price\") return SortFunction::SIMPLE_MOVING_AVERAGE;
    if (sort_function_str == \"Exponential Moving Average of Price\") return SortFunction::EXPONENTIAL_MOVING_AVERAGE;
    if (sort_function_str == \"Standard Deviation of Return\") return SortFunction::STANDARD_DEVIATION_RETURN;
    if (sort_function_str == \"Moving Average of Return\") return SortFunction::MOVING_AVERAGE_RETURN;
    if (sort_function_str == \"current price\") return SortFunction::CURRENT_PRICE;
    if (sort_function_str == \"Portfolio Return\") return SortFunction::PORTFOLIO_RETURN;
    
    throw SortNodeError(\"Invalid sort function: \" + sort_function_str);
}

SelectFunction SortNodeProcessor::parse_select_function(const std::string& select_function_str) {
    if (select_function_str == \"Top\") return SelectFunction::TOP;
    if (select_function_str == \"Bottom\") return SelectFunction::BOTTOM;
    
    throw SortNodeError(\"Invalid select function: \" + select_function_str);
}

std::pair<nlohmann::json, bool> SortNodeProcessor::get_branches(const StrategyNode& node) {
    // Check if node has branches property
    if (node.branches.contains(\"branches\")) {
        return std::make_pair(node.branches[\"branches\"], true);
    }
    
    // Otherwise use the branches directly
    return std::make_pair(node.branches, false);
}

} // namespace atlas", "original_text": "", "replace_all": false}]